File: clients/haskell/src/Codegame/ClientMessage.hs
module Codegame.ClientMessage where

import Prelude hiding (id)
import qualified Trans
import Trans (Trans)
import Data.Int
import Debugging.DebugCommand (DebugCommand)
import Model.Order (Order)

-- | Ask app to perform new debug command
data ClientMessageDebugMessage = ClientMessageDebugMessage {
    -- | Command to perform
    command :: DebugCommand }
    deriving Show

instance Trans ClientMessageDebugMessage where
    read = do
        command <- Trans.read
        return ClientMessageDebugMessage {
            command }
    
    write ClientMessageDebugMessage {
        command } = do
            Trans.write command

-- | Reply for ServerMessage::GetOrder
data ClientMessageOrderMessage = ClientMessageOrderMessage {
    -- | Player's order
    order :: Order }
    deriving Show

instance Trans ClientMessageOrderMessage where
    read = do
        order <- Trans.read
        return ClientMessageOrderMessage {
            order }
    
    write ClientMessageOrderMessage {
        order } = do
            Trans.write order

-- | Signifies finish of the debug update
data ClientMessageDebugUpdateDone = ClientMessageDebugUpdateDone { }
    deriving Show

instance Trans ClientMessageDebugUpdateDone where
    read = do
        return ClientMessageDebugUpdateDone { }
    
    write ClientMessageDebugUpdateDone { } = do
            return ()

-- | Request debug state from the app
data ClientMessageRequestDebugState = ClientMessageRequestDebugState { }
    deriving Show

instance Trans ClientMessageRequestDebugState where
    read = do
        return ClientMessageRequestDebugState { }
    
    write ClientMessageRequestDebugState { } = do
            return ()

-- | Message sent from client
data ClientMessage
    -- | Ask app to perform new debug command
    = DebugMessage ClientMessageDebugMessage
    -- | Reply for ServerMessage::GetOrder
    | OrderMessage ClientMessageOrderMessage
    -- | Signifies finish of the debug update
    | DebugUpdateDone ClientMessageDebugUpdateDone
    -- | Request debug state from the app
    | RequestDebugState ClientMessageRequestDebugState
    deriving Show

instance Trans ClientMessage where
    read = do
        tag :: Int32 <- Trans.read
        case tag of
            0 -> DebugMessage <$> Trans.read
            1 -> OrderMessage <$> Trans.read
            2 -> DebugUpdateDone <$> Trans.read
            3 -> RequestDebugState <$> Trans.read
    
    write (DebugMessage value) = do
        Trans.write (0 :: Int32)
        Trans.write value
    write (OrderMessage value) = do
        Trans.write (1 :: Int32)
        Trans.write value
    write (DebugUpdateDone value) = do
        Trans.write (2 :: Int32)
        Trans.write value
    write (RequestDebugState value) = do
        Trans.write (3 :: Int32)
        Trans.write value

File: clients/go/pkg/codegame/client_message.go
package codegame

import (
	"aicup22/pkg/debugging"
	"aicup22/pkg/model"
	"aicup22/pkg/stream"
)

// Message sent from client
type ClientMessage interface {
	// Write ClientMessage to writer
	Write()

	// Get string representation of ClientMessage
	String() string
}

// Read ClientMessage from reader
func ReadClientMessage() ClientMessage {
	switch stream.Flow().ReadInt32() {
	case 0:
		return ReadClientMessageDebugMessage()
	case 1:
		return ReadClientMessageOrderMessage()
	case 2:
		return ReadClientMessageDebugUpdateDone()
	case 3:
		return ReadClientMessageRequestDebugState()
	}
	// FIXME: НИКОГДА не используйте панику в инфраструктурном коде
	panic("Unexpected tag value")
}

// Ask app to perform new debug command
type ClientMessageDebugMessage struct {
	// Command to perform
	Command debugging.DebugCommand
}

func NewClientMessageDebugMessage(command debugging.DebugCommand) ClientMessageDebugMessage {
	return ClientMessageDebugMessage{
		Command: command,
	}
}

// Read DebugMessage from reader
func ReadClientMessageDebugMessage() ClientMessageDebugMessage {
	command := debugging.ReadDebugCommand()
	return ClientMessageDebugMessage{
		Command: command,
	}
}

// Write DebugMessage to writer
func (clientMessageDebugMessage ClientMessageDebugMessage) Write() {
	flow.WriteInt32(0)
	command := clientMessageDebugMessage.Command
	command.Write()
}

// Get string representation of DebugMessage
func (clientMessageDebugMessage ClientMessageDebugMessage) String() string {
	stringResult := "{ "
	stringResult += "Command: "
	command := clientMessageDebugMessage.Command
	stringResult += command.String()
	stringResult += " }"
	return stringResult
}

// Reply for ServerMessage::GetOrder
type ClientMessageOrderMessage struct {
	// Player's order
	Order model.Order
}

func NewClientMessageOrderMessage(order model.Order) ClientMessageOrderMessage {
	return ClientMessageOrderMessage{
		Order: order,
	}
}

// Read OrderMessage from reader
func ReadClientMessageOrderMessage() ClientMessageOrderMessage {
	order := model.ReadOrder()
	return ClientMessageOrderMessage{
		Order: order,
	}
}

// Write OrderMessage to writer
func (clientMessageOrderMessage ClientMessageOrderMessage) Write() {
	flow.WriteInt32(1)
	order := clientMessageOrderMessage.Order
	order.Write()
}

// Get string representation of OrderMessage
func (clientMessageOrderMessage ClientMessageOrderMessage) String() string {
	stringResult := "{ "
	stringResult += "Order: "
	order := clientMessageOrderMessage.Order
	stringResult += order.String()
	stringResult += " }"
	return stringResult
}

// Signifies finish of the debug update
type ClientMessageDebugUpdateDone struct {
}

func NewClientMessageDebugUpdateDone() ClientMessageDebugUpdateDone {
	return ClientMessageDebugUpdateDone{}
}

// Read DebugUpdateDone from reader
func ReadClientMessageDebugUpdateDone() ClientMessageDebugUpdateDone {
	return ClientMessageDebugUpdateDone{}
}

// Write DebugUpdateDone to writer
func (clientMessageDebugUpdateDone ClientMessageDebugUpdateDone) Write() {
	// FIXME: магическая константа
	flow.WriteInt32(2)
}

// Get string representation of DebugUpdateDone
func (clientMessageDebugUpdateDone ClientMessageDebugUpdateDone) String() string {
	stringResult := "{ "
	stringResult += " }"
	return stringResult
}

// Request debug state from the app
type ClientMessageRequestDebugState struct {
}

func NewClientMessageRequestDebugState() ClientMessageRequestDebugState {
	return ClientMessageRequestDebugState{}
}

// Read RequestDebugState from reader
func ReadClientMessageRequestDebugState() ClientMessageRequestDebugState {
	return ClientMessageRequestDebugState{}
}

// Write RequestDebugState to writer
func (clientMessageRequestDebugState ClientMessageRequestDebugState) Write() {
	flow.WriteInt32(3)
}

// Get string representation of RequestDebugState
func (clientMessageRequestDebugState ClientMessageRequestDebugState) String() string {
	stringResult := "{ "
	stringResult += " }"
	return stringResult
}
